design/automatic-network-configuration.md

```diff
@@ -0,0 +1,355 @@
+  - If a host has multiple interfaces, selects one for network configuration.
+  - Queries the configuration status of the SwitchPort CR to ensure that the
+    configuration is successful before provisioning.
+
+The main implementation process is described below:
+- After creating a BMH, the administrator puts an interface into the
+  provisioning network.
+
+  ![automatically configure 1](images/automatically_configure_1.png)
+
+- After an Interface is put into the provisioning network, the LLDP information
+  of other interfaces can be obtained and network configuration can be
+  performed.
+
+  ![automatically configure 2](images/automatically_configure_2.png)
+
+The workflow of a basic user story:
+- Administrator:
+  - Creates a Network resource used as provisioning network.
+  - Creates BMH CRs. Need to know the following information about BMH:
+    - Info of the interface which will be connected to the provisioning
+      network.
+    - Info of the switch port connected to this interface.
+  - Creates an Interface CR for each BMH based on the above information.
+    - The Interface controller monitors the Interface CR, finds the
+      corresponding SwitchPort CR according to the switch port information, and
+      fills it in `.spec.neighborRef` in the Interface CR.
+  - Adds the Interface CR information to `.spec.interfaces` in the provisioning
+    Network CR.
+    - The Network controller watches Network CRs and for each interfaces in its
+      `.spec.interfaces`, the controller will modify the corresponding
+      SwitchPort CR to match the network configuration.
+    - The Switch controller will configure the switch based on the SwitchPort
+      CR.
+    - After the switch configuration is completed, the BMH is connected to the
+      provisioning network. Ironic will perform introspection and update the
+      status of the BMH. The `.status.hardware.nics` in the BMH CR will contain
+      the LLDP information collected on each NICs.
+    - CAPM3 monitors BMH CRs and obtains the information of each interface
+      stored in `.status.hardware.nics`. CAPM3 will also create the
+      corresponding Interface CR if it is not created.
+- User:
+  - Creates production Network CRs.
+  - Creates a Cluster and specify its `.spec.networkRef`.
+    - CAPM3 finds a ready BMH.
+    - CAPM3 selects an Interface from the Interface CRs owned by this BMH to
+      connect to the network specified by the user.
+    - CAPM3 writes the chosen Interface CR information into the Network CR.
+    - Network Controller and Switch Controller will complete the automatic
+      configuration of the switch.
+    - After CAPM3 observes that the SwitchPort configuration is complete, it
+      triggers provisioning via the provisioning network.
+    - When the provisioning is completed, CAPM3 removes the bmh's Interface for
+      provisioning from the provisioning Network CR, thus disconnecting the BMH
+      from the provisioning network.
```

Comment:
Is this desired ? In some cases probably, but it should be up to the user. For example if we pivot, to have a self hosted cluster, the nodes that run ironic must have access to the provisioning network. so that should be up to the user.